tests/core/CoreTest.py

```diff
@@ -0,0 +1,184 @@
+
+        self.section1 = Section('test-section1')
+        self.filedict1 = {'f1': []}
+
+    @staticmethod
+    def execute_run(bears):
+        results = []
+
+        def on_result(result):
+            results.append(result)
+
+        run(bears, on_result)
+
+        return results
+
+    @staticmethod
+    def get_comparable_results(results):
+        """
+        Transforms an iterable of ``TestResult`` into something comparable.
+
+        Some ``TestResult`` instances returned by ``run`` contain instance
+        values. Due to the ``ProcessPoolExecutor``, objects get pickled,
+        are transferred to the other process and are re-instantiated,
+        effectively changing the id of them. The same happens again on the
+        transfer back in the results, so we need something that can be
+        compared.
+
+        This function extracts relevant values into a tuple, containing::
+
+            (test_result.bear.name,
+             test_result.section_name,
+             test_result.file_dict)
+
+        :param results:
+            The results to transform.
+        :return:
+            A list of comparable results for tests.
+        """
+        return [(result.bear.name, result.section_name, result.file_dict)
+                for result in results]
+
+    def assertTestResultsEqual(self, real, expected):
+        """
+        Test whether results from ``execute_run`` do equal with the ones given.
+
+        This function does a sequence comparison without order, so for example
+        ``[1, 2, 1]`` and ``[2, 1, 1]`` are considered equal.
+
+        :param real:
+            The actual results.
+        :param expected:
+            The expected results.
+        """
+        comparable_real = self.get_comparable_results(real)
+
+        self.assertEqual(len(comparable_real), len(expected))
+        for result in expected:
+            self.assertIn(result, comparable_real)
+            comparable_real.remove(result)
+
+    def test_run_simple(self):
+        # Test single bear without dependencies.
+        bear = MultiTaskBear(self.section1, self.filedict1, tasks_count=1)
+
+        results = self.execute_run({bear})
+
+        self.assertEqual(results, [0])
+
+    def test_run_result_handler_exception(self):
+        # Test exception in result handler. The core needs to retry to invoke
+        # the handler and then exit correctly if no more results and bears are
+        # left.
+        bear = MultiTaskBear(self.section1, self.filedict1, tasks_count=10)
+
+        on_result = unittest.mock.Mock(side_effect=ValueError)
+
+        run({bear}, on_result)
+
+        on_result.assert_has_calls([unittest.mock.call(i) for i in range(10)],
+                                   any_order=True)
+
+    def test_run_bear_exception(self):
+        # Test exception in bear. Core needs to shutdown directly and not wait
+        # forever.
+        self.execute_run({FailingBear(self.section1, self.filedict1)})
```

Comment:
maybe I should also run another bear in another test case and see whether it gets executed properly although the other one crashed^^